File: solutions/python/algorithm/priorityqueue/Exercise4.py
from structure.dynamic.DynamicPriorityQueueADT import DynamicPriorityQueueADT


def filtrar_por_prioridad(cola, prioridad_minima):
    resultado = DynamicPriorityQueueADT()
    aux = DynamicPriorityQueueADT()

    while not cola.is_empty():
        valor = cola.get_element()
        prioridad = cola.get_priority()
        cola.remove()
        aux.add(valor, prioridad)

        if prioridad >= prioridad_minima:
            resultado.add(valor, prioridad)

    # Restaurar cola original
    while not aux.is_empty():
        valor = aux.get_element()
        prioridad = aux.get_priority()
        aux.remove()
        cola.add(valor, prioridad)

    return resultado


# Ejemplo de uso
cola = DynamicPriorityQueueADT()
cola.add(10, 2)
cola.add(20, 3)
cola.add(30, 2)
cola.add(40, 1)

resultado = filtrar_por_prioridad(cola, 3)

while not resultado.is_empty():
    print(f"Valor: {resultado.get_element()}, Prioridad: {resultado.get_priority()}")
    resultado.remove()

File: solutions/javascript/src/algorithm/priorityqueue/Exercise4.js
import DynamicPriorityQueueADT from "../../structure/dynamic/DynamicPriorityQueueADT.js";


function filtrarColaPorPrioridad(originalQueue, prioridadMinima) {
  const nuevaCola = new DynamicPriorityQueueADT();
  const auxiliar = new DynamicPriorityQueueADT();

  while (!originalQueue.isEmpty()) {
    const valor = originalQueue.getElement();
    const prioridad = originalQueue.getPriority();

    if (prioridad >= prioridadMinima) {
      nuevaCola.add(valor, prioridad);
    }

    auxiliar.add(valor, prioridad); // Para restaurar
    originalQueue.remove();
  }

  // Restaurar cola original
  while (!auxiliar.isEmpty()) {
    originalQueue.add(auxiliar.getElement(), auxiliar.getPriority());
    auxiliar.remove();
  }

  return nuevaCola;
}


const cola = new DynamicPriorityQueueADT();

cola.add(10, 1);
cola.add(20, 5);
cola.add(30, 3);
cola.add(40, 7);
cola.add(50, 2);

const prioridadMinima = 3;
const filtrada = filtrarColaPorPrioridad(cola, prioridadMinima);

console.log(`Elementos con prioridad >= ${prioridadMinima}:`);
while (!filtrada.isEmpty()) {
    console.log(`[${filtrada.getElement()}, ${filtrada.getPriority()}]`);
    filtrada.remove();
}

console.log("\nCola original después del filtrado (debe mantenerse igual):");
while (!cola.isEmpty()) {
    console.log(`[${cola.getElement()}, ${cola.getPriority()}]`);
    cola.remove();
}